authentication/src/test/java/io/camunda/authentication/filters/OAuth2RefreshTokenFilterTest.java

```diff
@@ -0,0 +1,464 @@
+    // Given
+    final OAuth2AuthorizedClient expiredClient = createExpiredAuthorizedClientWithRefreshToken();
+    when(securityContext.getAuthentication()).thenReturn(authenticationToken);
+    when(authorizedClientService.loadAuthorizedClient(CLIENT_REGISTRATION_ID, PRINCIPAL_NAME))
+        .thenReturn(expiredClient);
+    when(authorizedClientManager.authorize(any(OAuth2AuthorizeRequest.class)))
+        .thenThrow(new OAuth2AuthorizationException(new OAuth2Error("invalid_grant")));
+
+    // When & Then
+    assertThatThrownBy(() -> filter.doFilterInternal(request, response, filterChain))
+        .isInstanceOf(OAuth2AuthenticationException.class)
+        .hasMessageContaining("invalid_grant");
+
+    assertThat(authenticationToken.isAuthenticated()).isFalse();
+    verify(logoutHandler).logout(request, response, authenticationToken);
+  }
+
+  @Test
+  void shouldThrowExceptionWhenRefreshedTokenIsExpired() {
+    // Given
+    final OAuth2AuthorizedClient expiredClient = createExpiredAuthorizedClientWithRefreshToken();
+    final OAuth2AuthorizedClient stillExpiredClient =
+        createExpiredAuthorizedClientWithRefreshToken();
+
+    when(securityContext.getAuthentication()).thenReturn(authenticationToken);
+    when(authorizedClientService.loadAuthorizedClient(CLIENT_REGISTRATION_ID, PRINCIPAL_NAME))
+        .thenReturn(expiredClient);
+    when(authorizedClientManager.authorize(any(OAuth2AuthorizeRequest.class)))
+        .thenReturn(stillExpiredClient);
+
+    // When & Then
+    assertThatThrownBy(() -> filter.doFilterInternal(request, response, filterChain))
+        .isInstanceOf(OAuth2AuthenticationException.class)
+        .hasMessageContaining("Access token expired");
+
+    assertThat(authenticationToken.isAuthenticated()).isFalse();
+    verify(logoutHandler).logout(request, response, authenticationToken);
+  }
+
+  @Test
+  void shouldUseCachedTokenWhenAvailableAndNotExpired() throws ServletException, IOException {
+    // Given
+    final OAuth2AuthorizedClient expiredClient = createExpiredAuthorizedClientWithRefreshToken();
+    final OAuth2AuthorizedClient refreshedClient = createValidAuthorizedClient();
+
+    when(securityContext.getAuthentication()).thenReturn(authenticationToken);
+    when(authorizedClientService.loadAuthorizedClient(CLIENT_REGISTRATION_ID, PRINCIPAL_NAME))
+        .thenReturn(expiredClient);
+    when(authorizedClientManager.authorize(any(OAuth2AuthorizeRequest.class)))
+        .thenReturn(refreshedClient);
+
+    // first request should refresh and cache
+    filter.doFilterInternal(request, response, filterChain);
+
+    // second request should use cache
+    filter.doFilterInternal(request, response, filterChain);
+
+    // Then
+    verify(filterChain, times(2)).doFilter(request, response);
+    verify(authorizedClientManager, times(1)).authorize(any(OAuth2AuthorizeRequest.class));
+  }
+
+  @Test
+  void shouldHandleConcurrentRequestsForSamePrincipal() throws Exception {
```

Comment:
This test is not very robust, because the two threads are running completely independent of each other (e.g. if we see it fail in the future, it may be hard to reproduce). If we want to keep it, I suggest we synchronize the threads in their execution. For example, we first let `thread1` run until it is in the critical section (here we will need to add some logic to synchronize). Then we let `thread2` start so that it reaches the lock and doesn't proceed. Then we trigger `thread1` to continue to refresh the token and complete. Accordingly, `thread1` continues and doesn't refresh the token.

It's possible to write this test but also a bit tricky. I could help with that. Let's discuss first if we need this synchronization in the implementation.